c2rust-analyze/src/panic_detail.rs

```diff
@@ -0,0 +1,205 @@
+    backtrace: Option<Backtrace>,
+    span: Span,
+}
+
+impl PanicDetail {
+    /// Create a new `PanicDetail` containing only a message, with no location or backtrace
+    /// information.
+    pub fn new(msg: String) -> PanicDetail {
+        PanicDetail {
+            msg,
+            loc: None,
+            relevant_loc: None,
+            backtrace: None,
+            span: DUMMY_SP,
+        }
+    }
+
+    /// Returns `true` if this `PanicDetail` contains a backtrace.
+    pub fn has_backtrace(&self) -> bool {
+        self.backtrace.is_some()
+    }
+
+    /// Return a short (usually one-line) description of this panic.
+    pub fn to_string_short(&self) -> String {
+        let loc_str = self
+            .relevant_loc
+            .as_ref()
+            .or(self.loc.as_ref())
+            .map_or("[unknown]", |s| &*s);
+        format!("{}: {}", loc_str, self.msg.trim())
+    }
+
+    /// Return a full description of this panic, including a complete backtrace if available.
+    pub fn to_string_full(&self) -> String {
+        let mut s = String::new();
+        let loc_str = self.loc.as_ref().map_or("[unknown]", |s| &*s);
+        writeln!(s, "panic at {}: {}", loc_str, self.msg).unwrap();
+        if let Some(ref relevant_loc) = self.relevant_loc {
+            writeln!(s, "related location: {}", relevant_loc).unwrap();
+        }
+        if !self.span.is_dummy() {
+            writeln!(s, "source location: {:?}", self.span).unwrap();
+        }
+        if let Some(ref bt) = self.backtrace {
+            writeln!(s, "{:?}", bt).unwrap();
+        }
+        s
+    }
+}
+
+enum PanicState {
+    /// The current thread is not in a `panic_detail::catch_unwind` call.
+    OutsideCatchUnwind,
+    /// The current thread is inside a `panic_detail::catch_unwind` call, but no panic has occurred
+    /// yet.
+    InsideCatchUnwind,
+    /// The current thread panicked while inside `panic_detail::catch_unwind`, producing a
+    /// `PanicDetail`, and is in the process of unwinding.
+    Unwinding(PanicDetail),
+}
+
+thread_local! {
+    static CURRENT_PANIC_DETAIL: Cell<PanicState> =
+        Cell::new(PanicState::OutsideCatchUnwind);
+}
+
+/// Panic hook for use with [`std::panic::set_hook`].  This builds a `PanicDetail` for each panic
+/// and stores it for later retrieval by [`take_current`].
+pub fn panic_hook(default_hook: &dyn Fn(&PanicInfo), info: &PanicInfo) {
+    CURRENT_PANIC_DETAIL.with(|cell| {
+        // Take the old value, replacing it with something arbitrary.
+        let old = cell.replace(PanicState::OutsideCatchUnwind);
+        match old {
+            PanicState::OutsideCatchUnwind => {
+                // No special behavior is needed.  Call the default panic hook instead.
+                default_hook(info);
+                return;
+            }
+            PanicState::InsideCatchUnwind => {}
+            PanicState::Unwinding(pd) => {
+                warn!("discarding old panic detail: {:?}", pd);
```

Comment:
Would this happen if there's a panic during unwinding, for example?  In that case, it might be better to use the original `PanicDetail`, which would be the real error we care about, and just warn about a panic happening during unwinding.  Or maybe it's just best to abort here.

Reply:
Panicking during unwinding already triggers an immediate abort, so I don't think we need any special handling here.  This case should be impossible, so I'll change the `warn!` to a `panic!`.

Reply:
Oh, perfect!  That should work then.